pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
+                total_weight +=
+                    Self::process_ended_session(current_unprocessed_session, current_session);
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn process_ended_session(
+            unprocessed_session_id: SessionIndex,
+            active_session_id: SessionIndex,
+        ) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 3);
+            ActiveCollators::<T>::insert(
+                unprocessed_session_id,
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedVec::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            let minimum_sessions_required = T::MaxInactiveSessions::get() + 1;
+            if active_session_id >= minimum_sessions_required {
+                total_weight += T::DbWeight::get().writes(1);
+                let _ = <crate::pallet::ActiveCollators<T>>::remove(
+                    active_session_id - minimum_sessions_required,
+                );
+            }
+            <LastUnprocessedSession<T>>::put(active_session_id);
+            total_weight
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    if !active_collators.contains(&author) {
+                        total_weight += T::DbWeight::get().writes(1);
+                        active_collators
+                            .try_push(author)
+                            .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached)?;
+                    }
+                    Ok(())
+                },
+            );
+            total_weight
+        }
+    }
+}
+
+impl<T: Config> NodeActivityTrackingHelper<T::CollatorId> for Pallet<T> {
+    fn is_node_inactive(node: &T::CollatorId) -> bool {
+        if !<EnableInactivityTracking<T>>::get() {
+            return false;
+        }
+
+        let current_session = T::CurrentSessionIndex::session_index();
+
+        let minimum_sessions_required = T::MaxInactiveSessions::get();
+        if current_session < minimum_sessions_required {
+            return false;
+        }
+
+        for session_index in current_session.saturating_sub(T::MaxInactiveSessions::get())
+            ..current_session.saturating_sub(1u32)
+        {
+            if <ActiveCollators<T>>::get(session_index).contains(node) {
+                return false;
```

Comment:
is there a more "RUST" way to do this @tmpolaczyk @nanocryk ?

Reply:
Like this but it's ugly, for loop is fine

```rust
(current_session.saturating_sub(T::MaxInactiveSessions::get())..current_session.saturating_sub(1u32))
    .all(|session_index| !ActiveCollators::<T>::get(session_index).contains(node))
```

Reply:
Loop is fine